feat(ide): Crash Reporting

nkomonen-amazon opened the pull request:
## Problem:

An extension can cause the IDE to crash. We want to capture when the IDE crashes, at least being able
to give some sort of information (when possible) for when we think the IDE crashed.

## Solution:

Hack together a CrashReporting class that emits telemetry events when it detects that we have
crashed.

For details see the docstring of the CrashReporting class.

## Additional:

This is not confirmed to all work on windows yet, so it is disabled for windows currently.

<!---
    REMINDER:
    - Read CONTRIBUTING.md first.
    - Add test coverage for your changes.
    - Update the changelog using `npm run newChange`.
    - Link to related issues/commits.
    - Testing: how did you test your changes?
    - Screenshots
-->

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

nkomonen-amazon committed:
add doc about vscode behaviors

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

nkomonen-amazon committed:
dev: Command to force IDE ExtHost to crash

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

Hweinstock committed:
docs(ec2): add ec2 connect details to `arch_features.md` (#5652)

## Problem


## Solution


---

<!--- REMINDER: Ensure that your PR meets the guidelines in
CONTRIBUTING.md -->

License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

jpinkney-aws committed:
telemetry(amazonq): Add telemetry tracing for amazon q inline (#5655)

## Problem
- The flow from when a message is shown to the user and
rejected/accepted by the user isn't instrumented with any way to trace
events

## Solution
- Create a trace from when the event is shown (either by manual
invocation or auto trigger) and then connect all the telemetry events
through a traceId.

### Implementation details
- The traceId is initially created when manual invocation/auto trigger
occurs
- For every telemetry event thats in the function hierarchy of the
manual invocation/auto trigger, they automatically assume the same
traceId
- When an event listener is called it uses the traceId we've set in
TelemetryHelper to setup the next traceId. This is because the event no
longer belongs to the same context
- Inside of commands we use `telemetry.record` because it allows us to
override the default telemetry id that gets automatically created and
instead associate it with the known trace

lalexcaws committed:
feat(telemetry): include user intent for copy and insert code interactions (#5654)

## Problem
Need to gather acceptance metrics on generated unit tests

## Solution
Include user intent on copy and insert code interactions in telemetry to
form an acceptance rate

---

<!--- REMINDER: Ensure that your PR meets the guidelines in
CONTRIBUTING.md -->

License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

jpinkney-aws commented on packages/core/src/shared/crashReporting/crashReporting.ts:
```diff
@@ -0,0 +1,451 @@
+     */
+    public constructor(
+        dependencies = {
+            checkInterval: DevSettings.instance.get('crashReportInterval', 10000 * 60 * 5),
```
remove a 0 on the 10000

jpinkney-aws commented on packages/core/src/shared/telemetry/telemetryService.ts:
```diff
@@ -110,7 +110,10 @@ export class DefaultTelemetryService {
             const currTime = new globals.clock.Date()
             // This is noisy when running tests in vscode.
-            telemetry.session_end.emit({ value: currTime.getTime() - this.startTime.getTime(), result: 'Succeeded' })
+            telemetry.session_end.emit({
```
Ask Dhruvi if the value field is being used

jpinkney-aws committed:
telemetry(amazonq): measure e2e latency for amazon q requests (#5566)

## Problem:
- We can't measure e2e how long a chat request takes

## Solution:
- Plumb the trace id through to the webview -> vscode -> webview. This
allows us to know exactly what is happening to a message and how long
everything took to get there
- chat e2e latency can be linked with the other telemetry events via the
traceId

nkomonen-amazon committed:
feat(utils): Add some util functions for upcoming feature (#5657)

This adds some util functions that are needed for an upcoming feature.

Each commit is its own isolated util, so review them individually.

**A significant amount of the changes come from minor refactors in
files**

---

<!--- REMINDER: Ensure that your PR meets the guidelines in
CONTRIBUTING.md -->

License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

---------

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

ivikash committed:
fix(amazonq): verify path for .obsolete file (#5671)

verify path for .obsolete file


License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

justinmk3 committed:
refactor(test): toTextDocument, toTextEditor #5660

Problem:
Redundant code, suboptimal names and docstrings.

Solution:
- Introduce `toTextDocument` for tests such as `codeParsingUtil.test.ts`
  which don't need a TextEditor.
- Rename `openATextEditorWithText` => `toTextEditor`.

bweedop committed:
feat(amazonq): user can generate unit tests for selected code #5577

- Add command to Amazon Q to generate unit tests for selected code. The
  feature will be limited to internal Amazon users, initially.
- Add "Generate Tests" command.

bweedop committed:
revert(amazonq): Disable generate unit tests command and context menu for all users (#5674)

## Problem

Unit test generation feature is not ready for internal users yet. This
PR disables the feature while improvements are made.

nkomonen-amazon committed:
fix OS Utils Pc Session start method, should work for all OS

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

dhasani23 committed:
revert(amazonq): skip showing users unit test form (#5679)

## Problem

Revert https://github.com/aws/aws-toolkit-vscode/pull/5607 by not
showing users the skip tests form


## Solution

Above

justinmk3 committed:
fix(logging): logger functions do not handle varargs (#5680)

Problem:
Regression b2ebc10342f63 in the logger: `debug('%s %s', foo, bar')`
sends `foo, bar` as a single `[foo, bar]` list to the first `%s`.
Example:

2024-09-26 12:18:29.631 [debug] getClientId: determined clientId as: [
      '96be923b-a00d-46d4-add1-c877abfcb01c',
      undefined,
      '96be923b-a00d-46d4-add1-c877abfcb01c'
    ], process.env clientId was: { logID: 1 }, stored clientId was: %s

Solution:
Spread the args when passing to sendToLog().

Commit:
Release 1.27.0

Commit:
Release 3.26.0

Commit:
Update version to snapshot version: 3.27.0-SNAPSHOT

Commit:
Update version to snapshot version: 1.28.0-SNAPSHOT

Commit:
Merge release into master

tverney committed:
fix(amazonq /dev): unpredictable placement of LLM-created files #5672

## Problem
In VS Code, when user add 2 different folders in the workspace which
contain different root paths, RTS/LLM cannot decide where to place a
generated file.

## Solution
This fix add the default first folder selected as its default for
orphaned root created files, not blocking user to receive an empty patch
but also support better the workspace usage.

hayemaxi committed:
build: detach src.gen/ from monorepo (#5667)

Theory: dependabot is still trying to update src.gen/ dependencies
because they are managed by the root monorepo. So technically, there is
no way to avoid updating src.gen's package.json because the root
package.json is linked to it and impacted by dependabot changes for
subproject dependencies.

Let's make its own "subrepo" and see if we can get dependabot to ignore
it finally.

---

<!--- REMINDER: Ensure that your PR meets the guidelines in
CONTRIBUTING.md -->

License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

nkomonen-amazon committed:
feat: Crash Monitoring

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

redid the constructors + extracted state

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

hayemaxi committed:
build: remove package.json syncing (#5629)

Removes the need to copy/sync package JSON contents from core ->
subproject extensions on build and package.

Summary of changes:
- Delete syncPackageJson.ts and handlePackageJson.ts
- Consolidate testing scripts and make small improvements (e.g.
validation, handle multiple directories to import tests from)
- package.json is now fully complete in `amazonq/`, `toolkit/`. Changes
to extensions should be made only the respective file (except for
icons).
- `toolkit/` runs unit, e2e, and integ tests from `core/`. It will also
run any of its own tests (but those don't exist yet).
- Web tests now moved to and run from `amazonq/`. This was decided
because `amazonq/` is more likely to support web than `toolkit/` at this
time. Also, it didn't seem possible to run the web tests from `core/`
like the other tests because it relies on webpack and a separate test
runner.
- No longer create web-based webpacks for `core/`

Follow up work:
- Get [packageJSON
properties](https://github.com/aws/aws-toolkit-vscode/blob/46b409c3353ce0dafefb796f5994c1fa88a23b47/packages/core/src/shared/vscode/env.ts#L61-L74)
from `context.extension.packageJSON` instead.
- Cleanup any remaining unneeded tasks in `.vscode/tasks.json`
- Move icons out of `core/package.json`.
- Continue to investigate: can we run web tests from core like other
tests? Toolkit may have a use in web mode one day.

---

<!--- REMINDER: Ensure that your PR meets the guidelines in
CONTRIBUTING.md -->

License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

justinmk3 committed:
ci: snooze techdebt

bweedop committed:
feat(amazonq): Re-enable generate unit tests command and context menu item (#5699)

## Problem

Command and context menu item for unit test generation was disabled for
internal users before last release. We are ready to enable now.

## Solution

This PR enables the context menu item and commands for unit test
generation.

Hweinstock committed:
test(ec2): fix flaky windows test.  (#5661)

## Problem
The test case here:
https://github.com/aws/aws-toolkit-vscode/blob/359b1d8252a32e064482678110c5931630bd49e0/packages/core/src/test/awsService/ec2/model.test.ts#L132C9-L146C11

This test is unaware of the side-effect that it will attempt to
explicitly delete the keys when generating. Originally, the delete was
conditional on the keys existing in the directory, so this test case was
not a problem since they never existed. However, on windows that check
was faulty, leading to the case where the windows file system reported
that the keys existed, which lead it to attempting to delete the root
directory.

## Solution
- Modify `tryRun` to accept an `onStdOut` optional parameter that lets
us accept the overwrite prompt from `ssh-keygen`.
- Do not explicitly delete the files in the generation function, make
this part of the generation step.

The benefits of this approach are that we moved the side-effects to a
place where they are expected, in the key generation process itself. It
also inherently avoids the potentially faulty check if the key files
exist, as we only use `onStdOut` if when the keys exist.

### Notes
- `packages/core/src/awsService/ec2/sshKeyPair.ts:49` includes `any`
which should be `ChildProcess`. For some reason the import fails in CI,
so I left it as `any`. Refer to `attempt to import child process` commit
for logs.

---

<!--- REMINDER: Ensure that your PR meets the guidelines in
CONTRIBUTING.md -->

License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

---------

Co-authored-by: Justin M. Keyes <jmkeyes@amazon.com>

Hweinstock committed:
test(ec2): fix flaky tests regarding child node updates.  (#5698)

## Problem
The issue is here:
https://github.com/aws/aws-toolkit-vscode/blob/9eea1f3700930df79620ef26ff6a43450e593580/packages/core/src/awsService/ec2/explorer/ec2ParentNode.ts#L55-L64
This fires off multiple asynchronous calls on each child, with the
function returning before any are resolved. Thus, the child node could
be deleted from the set after this function returned (potentially when
another piece of code is using it). This leads to `undefined` errors
within that function.
## Solution
Refactor the code to make use of a `for...of` with `await` pattern. 

---

<!--- REMINDER: Ensure that your PR meets the guidelines in
CONTRIBUTING.md -->

License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

Hweinstock committed:
feat(filesystem): add stub for chmod to filesystem wrapper (#5647)

## Problem
Changing the permissions of a file is currently not supported by the
file system wrapper.

## Solution
Implement a stub chmod that does nothing on web, and uses `fs` to modify
permissions in non-web environments.

---

<!--- REMINDER: Ensure that your PR meets the guidelines in
CONTRIBUTING.md -->

License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

---------

Co-authored-by: Justin M. Keyes <jmkeyes@amazon.com>

Galaoaoa committed:
fix: rename Application Composer to Infrastructure Composer #5701

## Problem
We will rename Application Composer to Infrastructure Composer as our
service name. We need to do the same thing in VSCode toolkit

## Solution
We will rename all user visible things first

ivikash committed:
fix(auth): make view container responsive (#5700)

make auth container responsive

# Before - The content scales down to 0.6x when the width in less than
260px

https://github.com/user-attachments/assets/d38db517-4734-435f-8100-b6a65d502c75

# After
## Login Screen

https://github.com/user-attachments/assets/17b579e1-003f-42aa-967e-7565e66624f0

## Re-Auth Screen

https://github.com/user-attachments/assets/2cd8754d-6707-4cd4-a250-7b28a0cd7de5



License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

Jurredr committed:
feat(amazonq): changed legal text in prompt UI footer (#5691)

## Description
Updated the legal disclaimer text in the footer of the chat client. This
is done as a clarification of compliant use according to the AWS AI
policy.

nkomonen-amazon committed:
refactor(deps): remove unnecessary code (#5707)

We dont need this code anymore

---

<!--- REMINDER: Ensure that your PR meets the guidelines in
CONTRIBUTING.md -->

License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

justinmk3 committed:
Update packages/core/src/dev/activation.ts

nkomonen-amazon committed:
move constants to the main constant file

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

nkomonen-amazon committed:
extracted out of topic folder

no longer needed that we have a single file for crash reporting

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

nkomonen-amazon committed:
refactor(urls): Remove isCloud9 from constants file

The isCloud9() function was causing issues in the constants file.
The constants file should be a minimal module with as little external
dependencies as possible.

We had some strings that needed to resolve a url depending on c9 vs toolkit.

Solution:

Move the resolution of c9 vs toolkit docs to an external function. Update uses
everywhere.

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

nkomonen-amazon committed:
rename variable

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

nkomonen-amazon committed:
move file

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>
